readme.md

```diff
@@ -677,6 +677,45 @@ level1.isDestroyed() //True;
 level2.isDestroyed() //True;
 ```
 
+# Interception
+
+Needle provides support for interception of construction using `interceptors`.  Interceptors provide the ability for developers to hook into a types construction both immediately before and after a type is instanced.  This technique is useful when you need to configure an instance before that instance is injected into downstream consumers. For example, if we had a Car type which injected an Engine, we may wish to call the engine.tune() function before giving to the car instance.  Interceptors are considered global inside of needle.  Therefore if you create multiple scopes each instance being constructed will pass through the same set of interceptors.  
+
+## Creating an interceptor
+
+To create an interceptor is simple, we simply implement an interface called `IConstructionInterceptor` in our class.  The interface is generic and there we can provide the Type that we wish to target as a generic param.  Below is an example interceptor which implements the interface for the Engine type.  
+
+```typescript
+export class EngineInterceptor implements IConstructionInterceptor<typeof Engine> {
+    public target: typeof Engine = Engine;
+    public beforeCreate(context: IInjectionContext<typeof Engine>): void {
+        console.log(context);
+    }
+    public afterCreate(instance: Engine, context: IInjectionContext<typeof Engine>): void {
+        console.log(instance);
+        console.log(context);
+    }
+}
+```
+
+The interface requires we implement 3 members
+
+* `target` -  The type we wish to intercept
+* `beforeCreate` - A method that will be invoked directly before the type instanced and after all its constructor args are resolved. 
+* `afterCreate` -  A method that will be invoked immediately after the target type was instanced.  
+
+Each method call will receive an `injection context` object which provides information about the context of injection. This includes information such as the injector instance resolving the type, the configuration used during construction and an array of constructor args.
+
+## Registering an interceptor
+
+The easiest way to register an interceptor is to simply decorate your class with `@Interceptor()`.  This will automatically register your interceptor with the root injector.  However if you want to use the injector API you can also do as follows.  
```

Comment:
Is this enforced? Can you decorate a class that does not implement `IConstructionInterceptor` with `@Interceptor()`?

Reply:
When I try and constrain this I get complaints about the type not implementing the methods.  See below.  Removing the default generic value makes no diff.  

![image](https://user-images.githubusercontent.com/1564426/103358138-cc754580-4aac-11eb-8a35-0655402059ef.png)

Reply:
I think that I have tried to enforce what types can be decorated before and it's not that straightforward. I'll have a go in playground to see if I can get it to work.

Reply:
This seems to work:

```ts
interface IMyInterface{
    interestedIn: (constructor: new (...args: any[]) => any) => boolean;
}


export function Interceptor() {
    return (interceptor: new (...args: any[]) => IMyInterface) => {
    };
}

@Interceptor() // decoration works
class InterfaceImplementation implements IMyInterface{
    interestedIn(constructor: new (...args: any[]) => any): boolean {
        return false
    }
}

@Interceptor() // decoration does not work
class NonInterfaceImplementation{
}
```

[playground](https://www.typescriptlang.org/play?ts=3.8.3#code/FASwdgLgpgTgZgQwMZQAQEkCyBPdlaIoDewqZq40MUAztACZ4BcqAFEgPZh0wCuSEDjBZgoAdzYA6aQhgBzGiwRhsAbQC6ASlQBeAHypl2bftQAjDhwA2UZQG5gAX2AuoADwAOQiKji8wAiBcGPgwKB6CMKzaJOSo1BC8MGBslLDhkSLiUjLyioYqGiYGWLihhFDFqLHkjg7OwAACeFQZQtGoAPSdqPRQnDAIEEEpYkIA1jTASFYINDQhVBXoALYeNitQkEMjFGsbWxALpS0EyFA1ZGnUdFCMYOxcPPyZqKISrNKSsgpKhVq6AxGTQsCzWWwpS5xeJQRLJXwIKw0KCkWpOFzNUJtKLabq9fpCHbBegcWhvDg+MYwcbTWbzVAAOS4p3g51W6ygm22wy4JGcQA)

Reply:
This is not using generic params and then applying those to methods.  Not equal basically to what we have in the PR. Try here https://www.typescriptlang.org/play?ts=3.5.1#code/KYDwDg9gTgLgBDAnmYcBywDuBDARgG1QF44A7LOACgDpbsoBzAZwC45tTEBtAXQEo4RAHztOAbgBQE0JFhwAZgFdSAYxgBLCKTgBJUjGBQVwMDGiUBAbwlxbcKMBiKo2yuv2Hjp6Gx0BhLSYYKEU1TVI9AyMTMygBYThrOzsbOwBfSTSpaXBoeHco+WxjXT0AK2AwrQCPEBgAHgAVRuRUUANSABMmdCw8QkFRRBEkuwB6ACoJ1LsJuEaAC1QkFDhcYHcGOBVA4NCDTpnbCbGjh2xOrXxEBFa2ZtbJZMnp5Lg5xdR3CrVoBAX1D13EEOCVAewmEwICp1NgDggIBAju9TslzpdSNc4N9KrE2BxEE9xlNkR8ltstPJ1AxnHDwnBFExgJ04J1nJsKaQgiEqqRSai7OirjcdqQqTSoHStPjxEcXqS4DVufs-vQaQBbYD6HqM5kKP6i5W8-lnYAXYWco2xACCjFYit2PNiAAV6NhNVEmE0WighJkpPL3nAAELYJmWva-KDY0g-WIxqJeeMFQxFYw2OYAAU6wCYKig6lM9KVkfjkU8MWgPQtVIccEwS20KmK5PVgKZhxROVk+Q8UDTqH8jv24XL0W8UG9asccHaWu6vRwBGIQxGcpJbzJy1aCGn8PkfxT440WhNaLNGKxMD390ae8k69eyTmegAbhAANZ69YHuswckrF8QJcteqjfoo8DYPIUTYjAPSGqWqqMIomranACzYK+qDrFq9i5hA+BYZ2z5wJmYBuuqnIGHUDq1JB8CQAUcAQPIEZOuEZ52D+0DAH45wGJQorUTAvjlLi4Q1MJU6MI4Qh8Gw77qJ0D7PBuJFvp+eqdOoDhqFiUEwf+26rBhPQ4dowKgcYxGzKR5GSpRlmgqgnwxiCYHMaxRlwAw6hYdogEKmRFFUe0tHCewDEQExLFsSOp6bgKtgGYYfFmgJTlgb4IHOT6wDSQwskADShXUomxuJ1RaFJzR7nJCnRcpEhZBIKj4GGPQAKKkL55CJC1mZjkm5h8N2eTbO1kJwN1vXAENlbRuo6pgIQqFwboJbsVo80TvUgGxTN7jAGuyRgIoBDqCoazAL+c2kO+zYntoJDkJgcC2pKiD1DoYm8pJ7R7a0B09UdQhCBYRK2GdF1XSlUAaY99IvRQH3YF9lhuVZwBsId5BiKVImlBVf3VQD+2sbjx1wGk4N8CpdjQ-gl27jJhPk9NIN9SQlOQ3AjPM9xDhpXCwCCaTZVE3GEniw07OU-VcCKSyoxvP8gLUILd0PVKpDUGdTALGLdF00cWSnedTOw9BqX8aLmXGDjnPACVQntOVUtVXRgMoMDs0K0riTInY-7q3DCM63rjKGxj9vOwT1Mm8kWRZEAA

Reply:
Ok I need to rework this then.  As I need an instance to be registered.

Reply:
Ok fixes in and docs updated